include/ringmesh/basic/frame.h

```diff
@@ -0,0 +1,182 @@
+ *     * Redistributions of source code must retain the above copyright
+ *       notice, this list of conditions and the following disclaimer.
+ *     * Redistributions in binary form must reproduce the above copyright
+ *       notice, this list of conditions and the following disclaimer in the
+ *       documentation and/or other materials provided with the distribution.
+ *     * Neither the name of ASGA nor the
+ *       names of its contributors may be used to endorse or promote products
+ *       derived from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL ASGA BE LIABLE FOR ANY DIRECT,
+ * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
+ * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ *     http://www.ring-team.org
+ *
+ *     RING Project
+ *     Ecole Nationale Superieure de Geologie - GeoRessources
+ *     2 Rue du Doyen Marcel Roubault - TSA 70605
+ *     54518 VANDOEUVRE-LES-NANCY
+ *     FRANCE
+ */
+
+#pragma once
+
+#include <ringmesh/basic/common.h>
+
+/*!
+ * @file Basic classes for representing frames
+ * @author Pierre Anquez
+ */
+
+namespace RINGMesh
+{
+    namespace Geometry
+    {
+        struct Plane;
+        FORWARD_DECLARATION_DIMENSION_STRUCT( Line );
+        ALIAS_2D( Line );
+    }
+}
+
+namespace RINGMesh
+{
+    template < index_t DIMENSION >
+    class RINGMESH_API FrameBase
+    {
+    public:
+        virtual ~FrameBase();
+
+        const vecn< DIMENSION >& operator[]( index_t coord ) const
+        {
+            ringmesh_assert( coord < DIMENSION );
+            return axis_[coord];
+        }
+
+        vecn< DIMENSION >& operator[]( index_t coord )
+        {
+            ringmesh_assert( coord < DIMENSION );
+            return axis_[coord];
+        }
+
+    protected:
+        // Default Frame aligned on space axis
+        FrameBase()
+        {
+            for( auto coord : RINGMesh::range( DIMENSION ) )
+            {
+                axis_[coord][coord] = 1.;
+            }
+        }
+
+    private:
+        std::vector< vecn< DIMENSION > > axis_{ DIMENSION,
+            vecn< DIMENSION >() };
+    };
+    ALIAS_2D_AND_3D( FrameBase );
+
+    template < index_t DIMENSION >
+    class RINGMESH_API Frame : public FrameBase< DIMENSION >
+    {
+    public:
+        Frame() = default;
```

Comment:
If you want to have only Frame> 2 > and Frame< 3 >, you could use delete instead.

Reply:
I do not want that

Reply:
Then you have to provide the entire templated implementation in the header, so the compiler could generate the corresponding code.

Reply:
Ok for me